File: CSharp/99beers.cs
using System;

namespace CSharp
{
    class Program
    {
        static int bottles = 99;

        static int random()
        {
            var rand = new Random();
            return rand.Next(0, 101);
        }

        static void Main()
        {
            while (bottles != 0)
            {
                bool ifWaste = false;
                while (1 < bottles || bottles <= -1)
                {
                    verse (bottles);
                    bottles--;
                }

                Console.WriteLine($"{bottles} bottle of beer on the wall,");
                Console.WriteLine($"{bottles} bottle of beer.");
                if (random() > 80)
                {
                    Console.Write("If that one bottle should happens to ");
                    Console.WriteLine("fall,");
                    if (random() < 20)
                    {
                        ifWaste = true;
                    }
                }
                else
                {
                    Console.WriteLine("Take one down and pass it around.");
                }

                bottles = ending(bottles, ifWaste);
            }
        }

        static void verse(int bottles)
        {
            Console.WriteLine($"{bottles} bottle of beer on the wall,");
            Console.WriteLine($"{bottles} bottle of beer.");
            if (random() > 80)
            {
                Console.Write("If one of those bottles should happen to");
                Console.WriteLine("fall,");
                Console.Write($"{bottles - 1} bottles of beer on the");
                Console.WriteLine("wall.\n");
            }
            else if (random() < 20)
            {
                Console.WriteLine("If that one bottle should happens to fall,");
                Console.WriteLine("what a waste of alcohol!\n");
            }
            else
            {
                Console.WriteLine("Take one down and pass it around.");
                Console.Write($"{bottles - 1} bottles of beer on the");
                Console.WriteLine("wall.\n");
            }
        }

        static int infiniteBottles()
        {
            verse(0);
            return -1;
        }

        static int ending(int bottles, bool waste)
        {
            if (random() == 100 && bottles == 1)
            {
                if (waste)
                {
                    Console.WriteLine("what a waste of alcohol!\n");
                }
                else
                {
                    Console.WriteLine("No more bottles of beer on the wall.\n");
                }

                bottles = infiniteBottles();
            }
            else
            {
                if (waste)
                {
                    Console.WriteLine("what a waste of alcohol!\n");
                }
                else
                {
                    Console.WriteLine("No more bottles of beer on the wall.\n");
                }

                Console.WriteLine("No more bottles of beer on the wall,");
                Console.WriteLine("no more bottles of beer.");
                if (random() <= 60)
                {
                    bottles = 0;
                    if (random() <= 10)
                    {
                        Console.WriteLine("There's nothing else to fall,");
                        Console.Write("because there's no more bottles of ");
                        Console.WriteLine("beer on the wall.\n");
                    }
                    else
                    {
                        Console.Write("We've taken them down and passed them");
                        Console.WriteLine("around;");
                        Console.WriteLine("now we're drunk and passed out!\n");
                    }
                }
                else
                {
                    bottles = 99;
                    Console.WriteLine("Go to the store and buy some more,");
                    Console.Write($"{bottles} bottles of beer on the");
                    Console.WriteLine("wall.\n");
                }
            }

            return bottles;
        }
    }
}

File: TypeScript/99beers.ts
let bottles: number = 99;

function random() {
  return Math.floor(Math.random() * (100 - 0 + 1) + 0);
}

function verse(bottles: number) {
  console.log(`${bottles} bottles of beer on the wall,`);
  console.log(`${bottles} bottles of beer.`);
  if (random() > 80) {
    console.log("If one of those bottles should happen to fall,");
    console.log(`${bottles - 1} bottles of beer on the wall.\n`);
  }
  else if (random() < 20) {
    console.log("If that one bottle should happens to fall,");
    console.log("what a waste of alcohol!\n");
  }
  else {
    console.log("Take one down and pass it around.");
    console.log(`${bottles - 1} bottles of beer on the wall.\n`);
  }
}

function infiniteBottles() {
  verse(0);
  return -1;
}

function ending(bottles: number, waste: boolean) {
  if (random() == 100 && bottles == 1) {
    if (waste) {
      console.log("what a waste of alcohol!\n");
    }
    else {
      console.log("0 bottles of beer on the wall.\n");
    }

    bottles = infiniteBottles();
  }
  else {
    if (waste) {
      console.log("what a waste of alcohol!\n");
    }
    else {
      console.log("No more bottles of beer on the wall.\n");
    }

    console.log("No more bottles of beer on the wall,");
    console.log("no more bottles of beer.");
    if (random() <= 60) {
      bottles = 0;
      if (random() <= 10) {
        console.log("There's nothing else to fall,");
        console.log("because there's no more bottles of beer on the wall.\n");
      }
      else {
        console.log("We've taken them down and passed them around;");
        console.log("now we're drunk and passed out!\n");
      }
    }
    else {
      bottles = 99;
      console.log("Go to the store and buy some more,");
      console.log(`${bottles} bottles of beer on the wall.\n`);
    }
  }

  return bottles;
}

while (bottles != 0) {
  let ifWaste: boolean = false;
  while (1 < bottles || bottles <= -1) {
    verse(bottles);
    bottles = bottles - 1;
  }

  console.log(`${bottles} bottle of beer on the wall,`);
  console.log(`${bottles} bottle of beer.`);
  if (random() > 80) {
    console.log("If that one bottle should happen to fall,")
    if (random() < 20) {
      ifWaste = true;
    }
  }
  else {
    console.log("Take one down and pass it around.");
  }

  bottles = ending(bottles, ifWaste);
}
